llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +440,140 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+using Operation = AIEBaseInstrInfo::AbstractVecOp::OperationType;
+using AbstractVecOp = AIEBaseInstrInfo::AbstractVecOp;
+using OptionalVecOp = std::optional<AbstractVecOp>;
+
+bool getGlobalValue(const MachineInstr *MI,
+                    SmallPtrSet<const Value *, 4> &GVSet,
+                    MachineRegisterInfo &MRI) {
+
+  MI = getDefIgnoringCopiesAndBitcasts(MI->getOperand(1).getReg(), MRI);
+
+  for (auto MO : MI->uses()) {
+    if (MO.isGlobal()) {
+      GVSet.insert(MO.getGlobal());
+      return true;
+    }
+  }
+  return false;
+}
+
+// Recognize BROADCAST (GLOBAL)
+bool visitVBroadcast(AbstractVecOp VBroadcast, const AIEBaseInstrInfo *TII,
+                     MachineRegisterInfo &MRI,
+                     SmallPtrSet<const Value *, 4> &GVSet) {
+
+  return getGlobalValue(MRI.getVRegDef(VBroadcast.ScalarSrcReg), GVSet, MRI);
+}
+
+// Recognize SELECT ((SELECT | BROADCAST), ((SELECT | BROADCAST)))
+bool visitVSelect(AbstractVecOp VSelect, const AIEBaseInstrInfo *TII,
+                  MachineRegisterInfo &MRI,
+                  SmallPtrSet<const Value *, 4> &GVSet) {
+
+  bool Success = true;
+  for (const Register Reg : VSelect.VectorSrcRegs) {
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MRI.getVRegDef(Reg));
+    // We don't recognize this! Here we cannot have anything different
+    // from SELECT or BROADCAST.
+    if (!VOp)
+      return false;
+    if (VOp->Type == Operation::SELECT)
+      Success &= visitVSelect(*VOp, TII, MRI, GVSet);
+    else if (VOp->Type == Operation::BROADCAST)
+      Success &= visitVBroadcast(*VOp, TII, MRI, GVSet);
+    else
+      return false;
+  }
+
+  return Success;
+}
+
+// Recognize ADD ((ADD | SELECT), Unknown)
+bool visitVAdd(const AbstractVecOp &VAdd, const AIEBaseInstrInfo *TII,
+               MachineRegisterInfo &MRI, SmallPtrSet<const Value *, 4> &GVSet) {
+
+  bool UnknownSide = false;
+  bool KnownSide = false;
+
+  for (const Register Reg : VAdd.VectorSrcRegs) {
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MRI.getVRegDef(Reg));
+
+    // We don't recognize this, but this represents the other side of the
+    // add (offset). It can be anything, but if it leads to out-of-bounds,
+    // we don't care bacause it is UB anyway.
+    if (!VOp) {
+      UnknownSide = true;
+      continue;
+    }
+    // The pointer-side of the add.
+    if (VOp->Type == Operation::ADD)
+      KnownSide = visitVAdd(*VOp, TII, MRI, GVSet);
+    else if (VOp->Type == Operation::SELECT)
+      KnownSide = visitVSelect(*VOp, TII, MRI, GVSet);
+    else
+      return false;
+  }
+  return UnknownSide /*offset side is found*/ &&
+         KnownSide /*broadcasted side is found*/;
```

Comment:
CHECK: KnownSide tells us that operand can be traced wholly to a set of objects. UnknownSide might be anything, but cannot introduce new objects without also introducing undefined behaviour. I have a feeling we only need to check KnownSide. Moreover, I would not use the term 'broadcasted side'. We may later generalize to other ways to construct a vector of object designations. 'object side' rhymes better with 'offset side'.

Reply:
Note: you couldn't directly assign to KnownSide anymore, instead do a sticky update if the visit function returns true.